src/jv_type_invalid.c

```diff
@@ -0,0 +1,68 @@
+#include <stdint.h>
+#include <stddef.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdarg.h>
+#include <limits.h>
+#include <math.h>
+
+#include "jv_type_private.h"
+
+/*
+ * Invalid objects, with optional error messages
+ */
+
+
+#define JVP_FLAGS_INVALID_MSG   JVP_MAKE_FLAGS(JV_KIND_INVALID, JVP_PAYLOAD_ALLOCATED)
+
+typedef struct {
+  jv_refcnt refcnt;
+  jv errmsg;
+} jvp_invalid;
+
+jv jv_invalid_with_msg(jv err) {
+  if (JVP_HAS_KIND(err, JV_KIND_NULL))
+    return JV_INVALID;
+  jvp_invalid* i = jv_mem_alloc(sizeof(jvp_invalid));
+  i->refcnt = JV_REFCNT_INIT;
+  i->errmsg = err;
+
+  jv x = {JVP_FLAGS_INVALID_MSG, 0, 0, 0, {&i->refcnt}};
+  return x;
+}
+
+jv jv_invalid() {
+  return JV_INVALID;
+}
+
+jv jv_invalid_get_msg(jv inv) {
+  assert(JVP_HAS_KIND(inv, JV_KIND_INVALID));
+
+  jv x;
+  if (JVP_HAS_FLAGS(inv, JVP_FLAGS_INVALID_MSG)) {
+    x = jv_copy(((jvp_invalid*)inv.u.ptr)->errmsg);
+  }
+  else {
+    x = jv_null();
+  }
+
+  jv_free(inv);
+  return x;
+}
+
+int jv_invalid_has_msg(jv inv) {
+  assert(JVP_HAS_KIND(inv, JV_KIND_INVALID));
+  int r = JVP_HAS_FLAGS(inv, JVP_FLAGS_INVALID_MSG);
+  jv_free(inv);
```

Comment:
This free changes the semantics of this function. Previously, it would not free `inv`, and was only freeing the extra copy on the `msg` performed by `jv_invalid_get_msg` (see https://github.com/stedolan/jq/pull/1752/files/bf7f1f2b8176597a04bf4a8c00243567d3c7c83d#diff-5c46af0444da42be404b9ff2acf24da2L124)

Reply:
Well, it's a change of implementation to follow the actual semantics of every single place this function was called, which was resulting a memory leak. 
Please check that in all places where it was called the parameter has been `jv_copy`ied.

This is exactly one of the things I have tried addressing in my API name refactoring. Today there is a big mess because one can't tell by the naming convention whether the function will be consuming the value or not.

At the same time, there is a general tendency to consume parameters in the calls, so instead of fixing all places where this function was called I have decided to change the implementation. And actually optimised it as well - previously it would construct a message just to see that it wasn't nil

Reply:
> Well, it's a change of implementation to follow the actual semantics of every single place this function was called, which was resulting a memory leak.
> Please check that in all places where it was called the parameter has been `jv_copy`ied

Ah, I hadn't realized (yes, more evidence of your point. 😃). This is fine, then.